src/EVM/Types.hs

```diff
@@ -1106,8 +1108,114 @@ data GenericOp a
   deriving (Show, Eq, Ord, Functor)
 
 
--- Function Selectors ------------------------------------------------------------------------------
+-- | A model for a buffer, either in it's compressed form (for storing parsed
+-- models from a solver), or as a bytestring (for presentation to users)
+data BufModel
+  = Comp CompressedBuf
+  | Flat ByteString
+  deriving (Eq)
+instance Show BufModel where
+  show (Comp c) = "Comp " <> show c
+  show (Flat b) = "Flat 0x" <> bsToHex b
+
+-- | This representation lets us store buffers of arbitrary length without
+-- exhausting the available memory, it closely matches the format used by
+-- smt-lib when returning models for arrays
+data CompressedBuf
+  = Base { byte :: Word8, length :: W256}
+  | Write { byte :: Word8, idx :: W256, next :: CompressedBuf }
+  deriving (Eq, Show)
+
+-- | a final post shrinking cex, buffers here are all represented as concrete bytestrings
+data SMTCex = SMTCex
+  { vars :: Map (Expr EWord) W256
+  , addrs :: Map (Expr EAddr) Addr
+  , buffers :: Map (Expr Buf) BufModel
+  , store :: Map (Expr EAddr) (Map W256 W256)
+  , blockContext :: Map (Expr EWord) W256
+  , txContext :: Map (Expr EWord) W256
+  }
+  deriving (Eq, Show)
+
+instance Semigroup SMTCex where
+  a <> b = SMTCex
+    { vars = a.vars <> b.vars
+    , addrs = a.addrs <> b.addrs
+    , buffers = a.buffers <> b.buffers
+    , store = a.store <> b.store
+    , blockContext = a.blockContext <> b.blockContext
+    , txContext = a.txContext <> b.txContext
+    }
+
+instance Monoid SMTCex where
+  mempty = SMTCex
+    { vars = mempty
+    , addrs = mempty
+    , buffers = mempty
+    , store = mempty
+    , blockContext = mempty
+    , txContext = mempty
+    }
+
+class GetUnknownStr a where
+    getUnknownStr :: a -> String
+
+instance GetUnknownStr String where
+    getUnknownStr = id
+
+instance GetUnknownStr (String, Expr End) where
+    getUnknownStr (s, _) = s
+
+data ProofResult a (b :: Type) where
+    Qed :: ProofResult a b
+    Cex :: a -> ProofResult a b
+    Unknown :: GetUnknownStr b => b -> ProofResult a b
+    Error :: String -> ProofResult a b
+
+instance (Show a, Show b) => Show (ProofResult a b) where
+  show Qed = "Qed"
+  show (Cex c) = "Cex " <> show c
+  show (Unknown u) = "Unknown " <> show u
+  show (Error e) = "Error: " <> e
 
+instance (Eq a, Eq b) => Eq (ProofResult a b) where
+  x == y = case (x, y) of
+    (Unknown u1, Unknown u2) -> u1 == u2
+    (Error a, Error b)       -> a == b
+    (Cex a, Cex b)           -> a == b
+    (Qed, Qed)               -> True
+    _                        -> False
+
+type VerifyResult = ProofResult (Expr End, SMTCex) (String, Expr End)
+type EquivResult = ProofResult (SMTCex) String
+type SMT2Result = ProofResult (SMTCex) String
```

Comment:
I think the number two in there could be misleading a bit.  I don't see a need for it.
```suggestion
type SMTResult = ProofResult (SMTCex) String
```

Reply:
I would personally prefer to have `Sat` and `Unsat` as the variants of `SMTResult`, though I understand that for `VerifyResult` and `EquivResult` the current terminology makes more sense.

I see SMT and symbolic execution/equivalence checking as working on two different levels, so for me it would be natural to have different representation for them and then some simple translation between the layers.